### PR TITLE
Split out OTLP gRPC vs. HTTP endpoint configuration

### PR DESCRIPTION
**Pull request by alanwest**
Here's a can of worms some folks may have hoped would just go away 😄.

## Background

The specification of the endpoint configuration for the OTLP exporter has caused some confusion among users due to inconsistencies with the implementations of the language SDKs as well as the collector.

### Related issues
* https://github.com/open-telemetry/opentelemetry-specification/issues/1426
* https://github.com/open-telemetry/opentelemetry-collector/issues/2539

### Prior context

In a prior version of the specification, the endpoint configuration did not mandate it be a URL with scheme included. There also used to be an `insecure` configuration flag which was removed. See #1234 where these changes were made. It seems that these changes were originally intended for the benefit of making the configuration of OTLP/HTTP exporters more natural, yet the discussion on this PR seems to suggest that folks accepted that the changes would equally apply to OTLP/gRPC exporters.

### Existing implementations

I have studied a few of the OTLP/gRPC exporter implementations. Please correct me on any of the observations I have made - I am not an expert in all of these languages.

#### .NET and Java SDKs
* Requires a scheme in the endpoint configuration.
* Requires the scheme be `http` or `https`.
* Uses scheme to infer secure vs. insecure channel.
* Does not have an `insecure` configuration option.
* References: [.NET](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry.Exporter.OpenTelemetryProtocol/OtlpTraceExporter.cs#L80) and [Java](https://github.com/open-telemetry/opentelemetry-java/blob/43f1ecf7ede49104c2aac1c9e5448b7cf7859ce7/exporters/otlp/trace/src/main/java/io/opentelemetry/exporter/otlp/trace/OtlpGrpcSpanExporterBuilder.java#L88)

#### Go SDK and Collector
* A scheme is not permitted in the endpoint configuration. An error is thrown to the effect of `address https://some-host:4317/: too many colons in address`
  * Go has implemented handling of a scheme specifically for configuration from the `OTEL_EXPORTER_OTLP_ENDPOINT` environment variable. See https://github.com/open-telemetry/opentelemetry-go/issues/1885.
* Has the `insecure` configuration option no longer present in the specification.

#### JS/Node
* Allows for a scheme in the endpoint configuration.
* Requires the scheme, if present, be `http`, `https`, `grpc`, or `grpcs`. [Code here](https://github.com/open-telemetry/opentelemetry-js/blob/7fa4ff70d1aad5875cb14c6bff99c838228447bb/packages/opentelemetry-exporter-collector-grpc/src/util.ts#L117-L121)
* The scheme has no bearing on whether a secure or insecure channel will be used. Instead it appears that this is configured by instantiating and using a [`ChannelCredentials`](https://grpc.github.io/grpc/node/grpc.credentials.html). The `ChannelCredentials` are used [here](https://github.com/open-telemetry/opentelemetry-js/blob/7fa4ff70d1aad5875cb14c6bff99c838228447bb/packages/opentelemetry-exporter-collector-grpc/src/util.ts#L36-L37).

### Input from Collector SIG

Based on a conversation on this topic with folks (@tigrannajaryan @bogdandrutu @alolita) during the 5/26 collector SIG meeting, I am drafting this PR to reopen a discussion for how (and if) we wish to resolve these inconsistencies.

To summarize the discussion from 5/26 collector SIG meeting:

  > If the collector adopts support for a scheme being present in the endpoint configuration, then people would prefer that this be implemented in a  backwards compatible way that would maintain support for a scheme-less endpoint.

## Changes

In an attempt to address 1) the desire for backwards compatibility and 2) the confusion brought on from the inconsistencies in the various implementations, here's the idea I'm toying with:

1. Use different language for the OTLP/gRPC and OTLP/HTTP exporters for the endpoint configuration.
2. Re-introduce the `insecure` flag only for use by the OTLP/gRPC exporter and only when a scheme is not present in the endpoint configuration.

### CODE REVIEW

**Review by tigrannajaryan — APPROVED** [review]
Thanks @alanwest 

Please also add a changelog entry.

I believe this is a non-breaking change since it makes Endpoint processing more _permissive_. Every endpoint value that was acceptable previously is still acceptable. Some values which were previously not accepted and would result in an error will be now accepted. I think we can treat this as a non-breaking change.

File: specification/protocol/exporter.md [hunk 1]
```diff
@@ -10,7 +10,9 @@ The following configuration options MUST be available to configure the OTLP expo
...
+| Endpoint (OTLP/gRPC) | Target to which the exporter is going to send spans or metrics. The endpoint MUST accept a URL with or without a scheme. If it includes a scheme (http or https), a scheme of https indicates a secure connection. | `https://localhost:4317` | `OTEL_EXPORTER_OTLP_ENDPOINT` `OTEL_EXPORTER_OTLP_TRACES_ENDPOINT` `OTEL_EXPORTER_OTLP_METRICS_ENDPOINT` |
+| Insecure             | Whether to enable client transport security for the exporter's `grpc` connection. Only applies to OTLP/gRPC when a scheme is not included in the `endpoint`. OTLP/HTTP always uses the scheme provided for the endpoint. | `false` | `OTEL_EXPORTER_OTLP_INSECURE` `OTEL_EXPORTER_OTLP_SPAN_INSECURE` `OTEL_EXPORTER_OTLP_METRIC_INSECURE` |
```
Comment (anuraaga) [inline specification/protocol/exporter.md:15]: Does this mean SDKs all have to support this variable? I'd prefer not to support it - we already have a method for picking secure / insecure in the endpoint scheme, having another way of doing the same thing is extra documentation / support overhead (not to mention bringing back this problematic word `insecure` which security scanners don't like much). If the idea is to allow either (no-scheme url + insecure flag) or (scheme url) than that's ok but don't see that mentioned in this wording.

Comment (anuraaga) [reply specification/protocol/exporter.md:15]: Another point is that it's always been quite mysterious when `host:port` defaults to TLS - I don't think that matches most users' expectations, it's just a security push by us I think. The schemed URLs solved this problem nicely, if allowing non-schemed URLs I really recommend the default being insecure, I think it matches most users' expectations (even `curl google.com` will hit HTTP, not HTTPS, this is what people think IMO).

Comment (iNikem) [reply specification/protocol/exporter.md:15]: > like to maintain the ability to configure without a scheme. Go SDK also still supports the `insecure` config setting

Do they want to configure without scheme but still use TLS?

Comment (alanwest) [reply specification/protocol/exporter.md:15]: > Do they want to configure without scheme but still use TLS?

Correct, TLS is the default in Go.

Go looks something like this
```go
        driver := otlpgrpc.NewDriver(
		otlpgrpc.WithInsecure(), // This option is required if you do not want TLS otherwise TLS is the default
		otlpgrpc.WithEndpoint("localhost:4317"), // To my knowledge, including a scheme here does not work
	)
	exp, err := otlp.NewExporter(ctx, driver)
```

On the other hand, Go does support a scheme when configuration is done via the `OTEL_EXPORTER_OTLP_ENDPOINT` environment variable. The scheme `http` implies the `WithInsecure` option.

Comment (alanwest) [reply specification/protocol/exporter.md:15]: > SDKs will always have extra knobs compared to what's in the spec. Usually they're language specific but it doesn't seem terrible if one SDK supports insecure flag if they want to.

This stance seems reasonable.

Perhaps at a minimum do you think it makes sense to say for OTLP/gRPC:

1) SDKs MUST support endpoint with configured with a scheme and the scheme `https` indicates use TLS.
2) SDKs MAY support an endpoint without a scheme but if they do then an additional `insecure` configuration is required.

First point effectively maintains status quo. Second point enables the flexibility for SDKs and requires a level of consistency when adopting this.

Comment (anuraaga) [reply specification/protocol/exporter.md:15]: Yup I have no problem with text related to an `insecure` flag if it's optional.

Comment (alanwest) [reply specification/protocol/exporter.md]: I've clarified things so that implementing the `insecure` flag is optional if the SDK does not allow for an endpoint to be configured without a scheme.

**Review by anuraaga — APPROVED** [review]

**Comment by carlosalberto** [comment]
I feel like this needs more eyes. @open-telemetry/specs-approvers please review.

**Comment by carlosalberto** [comment]
@open-telemetry/specs-approvers Please do review/approve ;)

**Review by jmacd — APPROVED** [review]

File: specification/protocol/exporter.md [hunk 2]
```diff
@@ -10,7 +10,9 @@ The following configuration options MUST be available to configure the OTLP expo
...
+| Endpoint (OTLP/gRPC) | Target to which the exporter is going to send spans or metrics. The endpoint MUST accept a URL with a scheme and MAY accept a URL without a scheme. If it includes a scheme (http or https), a scheme of https indicates a secure connection. If it does not include a scheme then the `insecure` configuration option is used to determine whether a secure connection should be used. | `https://localhost:4317` | `OTEL_EXPORTER_OTLP_ENDPOINT` `OTEL_EXPORTER_OTLP_TRACES_ENDPOINT` `OTEL_EXPORTER_OTLP_METRICS_ENDPOINT` |
```
Comment (Aneurysm9) [inline specification/protocol/exporter.md:14]: For gRPC it is my understanding that the scheme is used to determine the address resolution method and not the transport protocol since the transport protocol is always HTTP/2, which requires TLS unless explicitly configured otherwise with and indication that insecure connections are acceptable.

[Valid scheme values for gRPC](https://github.com/grpc/grpc/blob/master/doc/naming.md) are `dns`, `unix`, `unix-abstract`, `ipv4`, and `ipv6`.  `dns` is the default value if no scheme is specified.

Using `http` and `https` as a shorthand for `dns-with-insecure` and `dns` may be acceptable, but we should be very explicit about what we're doing and how client implementations should treat otherwise invalid gRPC endpoint names.  We should also make clear that the other gRPC-supported schemes are valid.

File: specification/protocol/exporter.md [hunk 3]
```diff
@@ -10,7 +10,9 @@ The following configuration options MUST be available to configure the OTLP expo
...
+| Endpoint (OTLP/HTTP) | Target to which the exporter is going to send spans or metrics. The endpoint MUST be a valid URL with scheme (http or https) and host, and MAY contain a port and path. A scheme of https indicates a secure connection. When using `OTEL_EXPORTER_OTLP_ENDPOINT`, exporters SHOULD follow the collector convention of appending the version and signal to the path (e.g. `v1/traces` or `v1/metrics`). The per-signal endpoint configuration options take precedence and can be used to override this behavior. See the [OTLP Specification][otlphttp-req] for more details. | `https://localhost:4317` | `OTEL_EXPORTER_OTLP_ENDPOINT` `OTEL_EXPORTER_OTLP_TRACES_ENDPOINT` `OTEL_EXPORTER_OTLP_METRICS_ENDPOINT` |
```
Comment (iNikem) [inline specification/protocol/exporter.md:13]:
```suggestion
| Endpoint (OTLP/HTTP) | Target to which the exporter is going to send spans or metrics. The endpoint MUST be a valid URL with scheme (http or https) and host, and MAY contain a port and path. A scheme of https indicates a secure connection. When using `OTEL_EXPORTER_OTLP_ENDPOINT`, exporters SHOULD follow the collector convention of appending the version and signal to the path (e.g. `v1/traces` or `v1/metrics`), if not present already. The per-signal endpoint configuration options take precedence and can be used to override this behavior. See the [OTLP Specification][otlphttp-req] for more details. | `https://localhost:4317` | `OTEL_EXPORTER_OTLP_ENDPOINT` `OTEL_EXPORTER_OTLP_TRACES_ENDPOINT` `OTEL_EXPORTER_OTLP_METRICS_ENDPOINT` |
```

File: specification/protocol/exporter.md (hunk 2 above)
Comment (iNikem) [reply specification/protocol/exporter.md:14]: This is very important point IMO. If gRPC already assigns some meaning to schema part of their URLs then won't it be confusing if we offer completely different schemas with completely different meaning for "gRPC URLs"?

Comment (iNikem) [reply specification/protocol/exporter.md:14]: @alanwest You use the same environment variables for both /HTTP and /gRPC options. How should SDK  understand which protocol to use?

Comment (alanwest) [reply specification/protocol/exporter.md:14]: > You use the same environment variables for both /HTTP and /gRPC options. How should SDK understand which protocol to use?

I left them the same since it has always been the case that the same environment variables are to be used for both HTTP and gRPC exporters.

To my knowledge, some languages only support gRPC at the moment. For languages that support exporting via both gRPC and HTTP it seems common that these are shipped as two distinct components.

I do not know the history here, but the specification does mention a [`OTEL_EXPORTER_OTLP_PROTOCOL` environment variable](https://github.com/open-telemetry/opentelemetry-specification/blob/2a712f7626cb62b21ae210ef3175612e7744f285/specification/protocol/exporter.md#specify-protocol) with the note "We reserve the following environment variables for configuration of protocols in the future".

I imagine that the `OTEL_EXPORTER_OTLP_PROTOCOL` variable would be necessary if a language SDK decided to develop a single component that could be configured to export either by HTTP or gRPC.

Comment (alanwest) [reply specification/protocol/exporter.md]: > Using `http` and `https` as a shorthand for `dns-with-insecure` and `dns` may be acceptable, but we should be very explicit about what we're doing and how client implementations should treat otherwise invalid gRPC endpoint names. We should also make clear that the other gRPC-supported schemes are valid.

Agreed there's some room to make this more clear - I took a stab at new wording, but I'm open to any suggestions for making this any more clear. See [this commit](https://github.com/open-telemetry/opentelemetry-specification/pull/1729/commits/ecabd5580f62f315dca3c24da0bbccedfb6b8d58).

I've attempted to keep things as client implementation agnostic as possible yet striving to be very clear. Client implementations differ in the schemes they support. As a point of contrast to the [valid scheme values for gRPC](https://github.com/grpc/grpc/blob/master/doc/naming.md) you pointed out, the .NET gRPC library, for example, requires including a scheme of `http` or `https` - that is, it doesn't really have a notion of defaulting to the `dns` scheme and in fact if you explicitly add `dns://` it doesn't work.

That said, I think it's safe to require OTel SDKs to require the ability to declare `http(s)` - I think the risk is low that we would stomp on the semantics of any specific client implementation.

Comment (Aneurysm9) [reply specification/protocol/exporter.md]: This is much better, I think.  That said, the handling of `insecure` still feels weird.  None of the gRPC standard schemes convey any information about whether TLS enforcement is necessary.  It seems wrong to say that `insecure` should only be honored when no scheme is present.  The address resolution scheme and TLS enforcement are distinct concerns for gRPC (and even for HTTP where a service may initiate TLS with an untrusted certificate and the user wants to ignore that).

Comment (alanwest) [reply specification/protocol/exporter.md]: > That said, the handling of insecure still feels weird.

Yep agreed, excellent callout. I've cleaned up the wording for the `insecure` option [here](https://github.com/open-telemetry/opentelemetry-specification/pull/1729/commits/3ca03ae902fc1ba864a883225b39ccadb8ab015e).

I've attempted again to remain as flexible and non-prescriptive as possible given the variety of client implementations some of which do not have a need for a separate insecure setting.

**Comment by tigrannajaryan** [comment]
@alanwest please resolve the conflicts.

**Comment by carlosalberto** [comment]
This has been around for some time now, and no more reviews/issues were reporting, hence merging.
